src/app/WriteHandler.cpp

```diff
@@ -754,23 +758,90 @@ void WriteHandler::MoveToState(const State aTargetState)
     ChipLogDetail(DataManagement, "IM WH moving to [%s]", GetStateStr());
 }
 
+DataModel::ActionReturnStatus WriteHandler::CheckWriteAllowed(const Access::SubjectDescriptor & aSubject,
+                                                              const ConcreteDataAttributePath & aPath)
+{
+    // TODO: ordering is to check writability/existence BEFORE ACL and this seems wrong, however
+    //       existing unit tests (TC_AcessChecker.py) validate that we get UnsupportedWrite instead of UnsupportedAccess
+    //
+    //       This should likely be fixed in spec (probably already fixed by
+    //       https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9024)
+    //       and tests and implementation
+    //
+    //       Open issue that needs fixing: https://github.com/project-chip/connectedhomeip/issues/33735
+    std::optional<DataModel::AttributeEntry> attributeEntry;
+    DataModel::AttributeFinder finder(mDataModelProvider);
+
+    attributeEntry = finder.Find(aPath);
+
+    // if path is not valid, return a spec-compliant return code.
+    if (!attributeEntry.has_value())
+    {
+        return DataModel::ValidateClusterPath(mDataModelProvider, aPath, Status::UnsupportedAttribute);
+    }
+
+    // Allow writes on writable attributes only
+    VerifyOrReturnValue(attributeEntry->writePrivilege.has_value(), Status::UnsupportedWrite);
+
+    bool checkAcl = true;
+    if (mLastSuccessfullyWrittenPath.has_value())
+    {
+        // NOTE: explicit cast/check only for attribute path and nothing else.
+        //
+        //       In particular `request.path` is a DATA path (contains a list index)
```

Comment:
There is no `request.path`.  There is aPath.  And we could make its type be `ConcreteAttributePath`, and then we don't have this problem at all and can just use `operator==`, right?

Reply:
Yes, copy & paste got to me. Updated text.
I had previously casts as `ConcreteAttributePath` however that pulls in `::operator==` and that is larger code than a manual compare.

Reply:
Hmm.  It's a little odd that the operator== is larger than literally the same code, but if it is, we should document that codesize is the reason we are not using it here, or something.

But I am _really_ suprised that code is larger, and would love to understand why.

Reply:
I used NM and it said "operator==" takes space. But then I used bloaty and it said the code was smaller... <insert confused image here...>

So ok, I then just use the more readable code.

Reply:
Well, after letting this run on CI, it does seem like operator== takes more space. No idea why.